Shell/Shell.cpp

```diff
@@ -1359,6 +1360,69 @@ void Shell::custom_event(Core::CustomEvent& event)
             Core::EventLoop::current().post_event(*this, make<Core::CustomEvent>(ShellEventType::ReadLine));
         return;
     }
+
+    if (event.custom_type() == ChildEvent) {
+#ifdef SH_DEBUG
+        dbg() << "SIGCHLD!";
+#endif
+        auto& jobs = this->jobs;
+        Vector<u64> disowned_jobs;
+        // Workaround the fact that we can't receive *who* exactly changed state.
+        // The child might still be alive (and even running) when this signal is dispatched to us
+        // so just...repeat until we find a suitable child.
+        // This, of course, will mean that someone can send us a SIGCHILD and we'd be spinning here forever.
+        bool found_child = false;
+        for (auto& it : jobs) {
+            auto job_id = it.key;
+            auto& job = *it.value;
+            int wstatus = 0;
+#ifdef SH_DEBUG
+            dbgf("waitpid({}) = ...", job.pid());
+#endif
+            auto child_pid = waitpid(job.pid(), &wstatus, WNOHANG | WUNTRACED);
+#ifdef SH_DEBUG
+            dbgf("... = {} - {}", child_pid, wstatus);
+#endif
+
+            if (child_pid < 0) {
+                if (errno == ECHILD) {
+                    // The child process went away before we could process its death, just assume it exited all ok.
+                    // FIXME: This should never happen, the child should stay around until we do the waitpid above.
+                    dbg() << "Child process gone, cannot get exit code for " << job_id;
+                    child_pid = job.pid();
+                } else {
+                    ASSERT_NOT_REACHED();
+                }
+            }
+            if (child_pid == 0) {
+                // If the child existed, but wasn't dead.
+                continue;
+            }
+            if (child_pid == job.pid()) {
+                if (WIFSIGNALED(wstatus) && !WIFSTOPPED(wstatus)) {
+                    job.set_signalled(WTERMSIG(wstatus));
+                } else if (WIFEXITED(wstatus)) {
+                    job.set_has_exit(WEXITSTATUS(wstatus));
+                } else if (WIFSTOPPED(wstatus)) {
+                    job.unblock();
+                    job.set_is_suspended(true);
+                }
+                found_child = true;
+            }
+            if (job.should_be_disowned())
+                disowned_jobs.append(job_id);
+        }
+        for (auto job_id : disowned_jobs)
+            jobs.remove(job_id);
+
+#ifndef SH_DEBUG
+        if (!found_child) {
+            dbg() << "DIDN'T FIND A CHILD?!?!?!?!?!?!?!?";
+            dbg() << "RETRY!";
+            Core::EventLoop::current().post_event(*this, make<Core::CustomEvent>(ShellEventType::ChildEvent));
```

Comment:
This spins for quite a while (some 100ms) before the child actually dies and shows up as dead through the `waitpid` above - possible finalisation bug? or maybe I'm just misunderstanding how this is supposed to work

Reply:
Waiting for *any* child is the correct approach. In absence of `SA_SIGINFO` there is just no way to know which child it was, so if you're doing a blocking `waitpid` on a particular PID you might end up deadlocking.

How long it may take, that depends. First, all threads need to be finalized, once all threads are finalized the process is finalized. All that assumes that the `FinalizerTask` had a chance to run. Whether that takes a little longer or not also depends on how the process was terminated, e.g. it may be necessary to briefly wake blocked threads to clean up the kernel stack before being able to flag it for finalization.

Reply:
> Waiting for any child is the correct approach

True, but that sounds really goofy, anyone can send us a `SIGCHLD`, and then we'd saturate the event loop waiting on nonexistent children.
Perhaps I should coalesce the events to at least limit the scope of this to "until the next child changes state" :thinking:

> Finalisation

I don't know the exact sequence of events here, but this signal is emitted in https://github.com/SerenityOS/serenity/blob/a9335eea1c96d50140e46b8e0361e01b87c09ce2/Kernel/Process.cpp#L622 - just before the process' memory regions are cleared and it's marked "dead".
so the finaliser _is_ run, but the process dies much later - or at least, that's the impression I'm getting.

Reply:
> True, but that sounds really goofy, anyone can send us a SIGCHLD, and then we'd saturate the event loop waiting on nonexistent children.
It's not really. `waitpid` will just return `ECHILD` in that case, telling you that there is no terminated child. So yes, the event loop will briefly execute to find a terminated child, but since it won't it should go back without doing anything else.

Reply:
> I don't know the exact sequence of events here, but this signal is emitted in
> 
> https://github.com/SerenityOS/serenity/blob/a9335eea1c96d50140e46b8e0361e01b87c09ce2/Kernel/Process.cpp#L622
> - just before the process' memory regions are cleared and it's marked "dead".
> so the finaliser _is_ run, but the process dies much later - or at least, that's the impression I'm getting.

Correct, this line is executed from `Process::finalize`, which is after all threads are finalized. Once finalized, its "shell" hangs around until you reap it with `waitpid`. Once reaped, it will be cleaned up entirely and can no longer be found with `waitpid`.

Reply:
So then why the rather large delay between the signal being delivered and `waitpid` considering the process dead?

Reply:
Good question. I wonder if it's because `Thread::send_signal` doesn't seem to wake the `Thread` it's sending a signal to if `this != Thread::current()`.

Reply:
> Good question. I wonder if it's because `Thread::send_signal` doesn't seem to wake the `Thread` it's sending a signal to if `this != Thread::current()`.

`Scheduler::pick_next` does look at `Thread::has_unmasked_pending_signals`, but that probably needs to wait until the thread quantum (time slice) is depleted. But that shouldn't take 100ms.